### PR TITLE
optimize: add doc comments for Location fields

### PR DESCRIPTION
**Pull request by kortschak**
Please take a look.

Updates #1235.

### CODE REVIEW

**Comment by kortschak** [comment]
I have added documentation about size restrictions that reflects the allocation changes from #1238.

PTAL

File: optimize/minimize.go [hunk 1]
```diff
@@ -35,10 +35,21 @@ type Task struct {
...
+	X []float64
+	// F is the result of evaluating the function at X.
+	F float64
+	// Gradient holds the gradient of the function at X.
+	// The length of Gradient must match the length of X
+	// or be zero. If the capacity of Gradient is less
+	// than the length of X, a new slice will be allocated.
 	Gradient []float64
-	Hessian  *mat.SymDense
+	// Hessian holds the curvature of the function at X.
```
Comment (vladimir-ch) [inline optimize/minimize.go]: I would avoid mentioning curvature, there is a relation but it is not that the Hessian is the curvature.

Comment (kortschak) [reply optimize/minimize.go]: What would you prefer?

Comment (vladimir-ch) [reply optimize/minimize.go]: I would just say "Hessian holds the Hessian of the function at X". That's 5 Hessians on 3 lines but maybe it doesn't matter.

Comment (kortschak) [reply optimize/minimize.go]: Mmm. I was hoping for something a little more descriptive. "Hessian holds the second-order partial derivatives of the function at X."?

Comment (vladimir-ch) [reply optimize/minimize.go]: Yeah, that's fine too and better than curvature. Then also gradient could be equally documented as "Gradient holds the first-order partial derivatives of the function at X".

Comment (kortschak) [reply optimize/minimize.go]: Done. PTAL

**Pull request merged**